devolutions-crypto-wayk/src/bastion/key_exchange.rs

```diff
@@ -1,9 +1,50 @@
 use crate::bastion::Error;
 use uuid::Uuid;
 use x25519_dalek::EphemeralSecret;
-use zeroize::Zeroizing;
+use zeroize::{Zeroize, Zeroizing};
+
+pub use x25519_dalek::PublicKey;
+
+/// Wraps shared secret produced by key exchange
+pub struct SharedSecret([u8; 32]);
+
+impl SharedSecret {
+    /// Constructs `SharedSecret` from a slice of bytes whose len is exactly 32.
+    /// Input slice is zeroized.
+    pub fn from_slice_mut(bytes: &mut [u8]) -> Result<Self, Error> {
+        use core::convert::TryFrom;
+        let array =
+            <[u8; 32]>::try_from(&*bytes).map_err(|_| Error::InvalidSize { got: bytes.len() })?;
+        bytes.zeroize();
+        Ok(Self(array))
+    }
+
+    pub fn from_bytes(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+
+    pub fn as_bytes(&self) -> &[u8; 32] {
+        &self.0
+    }
+}
+
+impl Zeroize for SharedSecret {
+    fn zeroize(&mut self) {
+        self.0.zeroize();
+    }
+}
 
-pub use x25519_dalek::{PublicKey, SharedSecret};
+impl Drop for SharedSecret {
+    fn drop(&mut self) {
+        self.zeroize();
+    }
+}
```

Comment:
This can be derived instead:

#[derive(Zeroize)]
#[zeroize(drop)]
pub struct SharedSecret([u8; 32]);